src/js/player.js

```diff
@@ -823,6 +825,213 @@ class Player extends Component {
+      set: (v) => {
+        const retval = srcDescriptor.set.call(el, v);
+
+        this.ready(() => this.trigger('sourceset'), true);
+
+        return retval;
+      },
+      configurable: true,
+      enumerable: srcDescriptor.enumerable
+    });
+
+    const oldSetAttribute = el.setAttribute;
+
+    el.setAttribute = (n, v) => {
+      const retval = oldSetAttribute.call(el, n, v);
+
+      if (n === 'src') {
+        this.ready(() => this.trigger('sourceset'), true);
+      }
+
+      return retval;
+    };
+
+    const oldLoad = el.load;
+
+    el.load = () => {
+      const retval = oldLoad.call(el);
+
+      this.ready(() => this.trigger('sourceset'), true);
+
+      return retval;
+    };
+
+    this.on('dispose', () => {
+      if (!el) {
+        return;
+      }
+
+      el.load = oldLoad.bind(el);
+      el.setAttribute = oldSetAttribute.bind(el);
+      Object.defineProperty(el, 'src', srcDescriptor);
+    });
+
+    // observer video tag changes, we do this when tech loads on iOS
+    this.on('sourceset', this.handleSourceSet_);
+  }
+
+  /**
+   * Update the internal source cache based on the `sourceset` event, and the resource selection
+   * algorithm.
+   *
+   * @see [Spec]{@link https://html.spec.whatwg.org/multipage/media.html#concept-media-load-algorithm}
+   *
+   * @param {EventTarget~Event} e
+   *        The sourceset event that triggered this function to run.
+   */
+  handleSourceSet_(e) {
+    const el = this.$('video') || this.$('audio');
+
+    if (!el) {
+      return;
+    }
+
+    // get what the video element thinks the source is
+    // and remove any querystring/hash from it
+    let src = el.src;
+    let sources = Array.prototype.map.call(el.getElementsByTagName('source'), (sourceEl) => {
+      return {src: sourceEl.src, type: sourceEl.type};
+    });
+
+    // if no source was set
+    if (!sources.length && typeof src === 'undefined') {
+      return;
+    }
+
+    // if we have el.src but there is no src attribute,
+    // el.src will be ignored, we should treat it like
+    // there is no el.src
+    if (el.getAttribute('src') === null) {
+      src = null;
+    }
+
+    // if we have a src we need to find out what the
+    // type for that source is
+    if (src) {
+      // get the source without any query string
+      const baseSrc = src.split(/[?#]/)[0];
+      let type;
+
+      // if we have sources see if we can
+      // find the type in those
+      if (sources) {
+        // get the source type from our sources list
+        for (let i = 0; i < sources.length; i++) {
+          const source = sources[i];
+          const baseSource = source.src.split(/[?#]/)[0];
+
+          if (source.src === src || baseSource === baseSrc) {
+            type = source.type;
+            break;
+          }
+        }
+      }
+
+      // otherwise we set the type to unknown,
+      // or in the case where we know the type based on the file extension
+      // we use that
+      if (!type) {
```

Comment:
As mentioned in the previous PR we need a better way to find the type here. Here are some of the questions I have:

* Should we reference a list of common source types and make it `unknown` when we don't know the type?
* Should we only update player.src?
* Should we use a `HEAD` request to get the `content-type` header for the source?